### PR TITLE
Redundant-array-inverse

### PR DESCRIPTION
**Pull request by alexnick83**
Added new RedundantArrayInverse transformation. Its functionality is similar to RedundantArray, but it removes instead the second access node. The transformation has been added to the list of strict transformations.

### CODE REVIEW

File: dace/transformation/dataflow/redundant_array.py [hunk 1]
```diff
@@ -97,4 +97,107 @@ def gnode(nname):
             RedundantArray._arrays_removed += 1
 
 
+class RedundantArrayInverse(pm.Transformation):
+    """ Implements the redundant array removal transformation, applied
+        when a transient array is copied from and to (from another array),
+        but never used anywhere else. This transformation remove the second
```
Comment (tbennun) [inline dace/transformation/dataflow/redundant_array.py]:
```suggestion
        but never used anywhere else. This transformation removes the second
```

File: dace/transformation/dataflow/redundant_array.py [hunk 2]
```diff
@@ -97,4 +97,107 @@ def gnode(nname):
             RedundantArray._arrays_removed += 1
 
 
+class RedundantArrayInverse(pm.Transformation):
```
Comment (tbennun) [inline dace/transformation/dataflow/redundant_array.py]: Can we change the name? The `Inverse` implies that there is some kind of inversion operation. What about `RedundantSecondArray`?

File: dace/transformation/dataflow/redundant_array.py [hunk 3]
```diff
@@ -97,4 +97,107 @@ def gnode(nname):
...
+    def expressions():
+        return [
+            nxutil.node_path_graph(RedundantArray._in_array,
+                                   RedundantArray._out_array)
+        ]
+
+    @staticmethod
+    def can_be_applied(graph, candidate, expr_index, sdfg, strict=False):
+        in_array = graph.nodes()[candidate[RedundantArray._in_array]]
+        out_array = graph.nodes()[candidate[RedundantArray._out_array]]
```
Comment (tbennun) [inline dace/transformation/dataflow/redundant_array.py]: Static variables must use the correct class here.

File: dace/transformation/dataflow/redundant_array.py [hunk 4]
```diff
@@ -97,4 +97,107 @@ def gnode(nname):
...
+        #         out_array.desc(sdfg).shape) or any(i != o for i, o in zip(
+        #             in_array.desc(sdfg).shape,
+        #             out_array.desc(sdfg).shape)):
+        #     return False
+
+        return True
+
+    @staticmethod
+    def match_to_str(graph, candidate):
+        out_array = graph.nodes()[candidate[RedundantArray._out_array]]
```
Comment (tbennun) [inline dace/transformation/dataflow/redundant_array.py]:
```suggestion
        out_array = graph.nodes()[candidate[RedundantArrayInverse._out_array]]
```

File: dace/transformation/dataflow/redundant_array.py [hunk 5]
```diff
@@ -97,4 +97,107 @@ def gnode(nname):
...
+
+        return "Remove " + str(out_array)
+
+    def apply(self, sdfg):
+        def gnode(nname):
+            return graph.nodes()[self.subgraph[nname]]
+
+        graph = sdfg.nodes()[self.state_id]
+        in_array = gnode(RedundantArray._in_array)
+        out_array = gnode(RedundantArray._out_array)
```
Comment (tbennun) [inline dace/transformation/dataflow/redundant_array.py]: Static variables must use the correct class here (too).

File: dace/transformation/dataflow/redundant_array.py [hunk 6]
```diff
@@ -97,4 +97,107 @@ def gnode(nname):
...
+                            pe.data.other_subset)
+
+            # Redirect edge to out_array
+            graph.remove_edge(e)
+            graph.add_edge(in_array, e.src_conn, e.dst, e.dst_conn, e.data)
+
+        # Finally, remove in_array node
+        graph.remove_node(out_array)
+        # TODO: Should the array be removed from the SDFG?
+        # del sdfg.arrays[in_array]
```
Comment (tbennun) [inline dace/transformation/dataflow/redundant_array.py]:
```suggestion
        # del sdfg.arrays[out_array]
```

File: dace/transformation/dataflow/redundant_array.py [hunk 7]
```diff
@@ -97,4 +97,107 @@ def gnode(nname):
...
+        for e in graph.out_edges(out_array):
+            # Modify all outgoing edges to point to in_array
+            path = graph.memlet_tree(e)
+            for pe in path:
+                if pe.data.data == out_array.data:
+                    pe.data.data = in_array.data
+                    if isinstance(subset, subsets.Indices):
+                        pe.data.subset.offset(subset, False)
+                    else:
+                        pe.data.subset = subset.compose(pe.data.subset)
```
Comment (tbennun) [inline dace/transformation/dataflow/redundant_array.py]: Nice! Forgot we had that function :)

File: dace/transformation/dataflow/redundant_array.py [hunk 8]
```diff
@@ -97,4 +97,107 @@ def gnode(nname):
...
+            # Redirect edge to out_array
+            graph.remove_edge(e)
+            graph.add_edge(in_array, e.src_conn, e.dst, e.dst_conn, e.data)
+
+        # Finally, remove in_array node
+        graph.remove_node(out_array)
+        # TODO: Should the array be removed from the SDFG?
+        # del sdfg.arrays[in_array]
+        if Config.get_bool("debugprint"):
+            RedundantArray._arrays_removed += 1
```
Comment (tbennun) [inline dace/transformation/dataflow/redundant_array.py]: This part can be removed. Also pointing to the wrong class.

Comment (alexnick83) [reply dace/transformation/dataflow/redundant_array.py]: Fixed

File: dace/transformation/dataflow/redundant_array.py (hunk 7 above)
Comment (alexnick83) [reply dace/transformation/dataflow/redundant_array.py]: Yes, seems to work.

File: dace/transformation/dataflow/redundant_array.py (hunk 6 above)
Comment (alexnick83) [reply dace/transformation/dataflow/redundant_array.py]: Fixed

File: dace/transformation/dataflow/redundant_array.py (hunk 5 above)
Comment (alexnick83) [reply dace/transformation/dataflow/redundant_array.py]: Fixed

File: dace/transformation/dataflow/redundant_array.py (hunk 4 above)
Comment (alexnick83) [reply dace/transformation/dataflow/redundant_array.py]: Fixed

File: dace/transformation/dataflow/redundant_array.py (hunk 3 above)
Comment (alexnick83) [reply dace/transformation/dataflow/redundant_array.py]: Fixed

File: dace/transformation/dataflow/redundant_array.py (hunk 2 above)
Comment (alexnick83) [reply dace/transformation/dataflow/redundant_array.py]: Fixed

File: dace/transformation/dataflow/redundant_array.py (hunk 1 above)
Comment (alexnick83) [reply dace/transformation/dataflow/redundant_array.py]: Fixed

**Pull request merged**